src/Interpreters/GraceHashJoin.cpp

```diff
@@ -0,0 +1,627 @@
+
+    size_t initial_num_buckets = roundUpToPowerOfTwoOrZero(std::clamp<size_t>(settings.grace_hash_join_initial_buckets, 1, settings.grace_hash_join_max_buckets));
+
+    for (size_t i = 0; i < initial_num_buckets; ++i)
+    {
+        addBucket(buckets);
+    }
+
+    if (buckets.empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "No buckets created");
+
+    LOG_TRACE(log, "Initialize {} buckets", buckets.size());
+
+    current_bucket = buckets.front().get();
+    current_bucket->startJoining();
+}
+
+bool GraceHashJoin::isSupported(const std::shared_ptr<TableJoin> & table_join)
+{
+    bool is_asof = (table_join->strictness() == JoinStrictness::Asof);
+    return !is_asof && isInnerOrLeft(table_join->kind()) && table_join->oneDisjunct();
+}
+
+GraceHashJoin::~GraceHashJoin() = default;
+
+bool GraceHashJoin::addJoinedBlock(const Block & block, bool /*check_limits*/)
+{
+    if (current_bucket == nullptr)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "GraceHashJoin is not initialized");
+
+    Block materialized = materializeBlock(block);
+    addJoinedBlockImpl(materialized);
+    return true;
+}
+
+bool GraceHashJoin::fitsInMemory() const
+{
+    /// One row can't be split, avoid loop
+    if (hash_join->getTotalRowCount() < 2)
+        return true;
+
+    return table_join->sizeLimits().softCheck(hash_join->getTotalRowCount(), hash_join->getTotalByteCount());
+}
+
+GraceHashJoin::Buckets GraceHashJoin::rehashBuckets(size_t to_size)
+{
+    std::unique_lock lock(rehash_mutex);
+    size_t current_size = buckets.size();
+
+    if (to_size <= current_size)
+        return buckets;
+
+    assert(isPowerOf2(to_size));
+
+    if (to_size > max_num_buckets)
+    {
+        throw Exception(ErrorCodes::LIMIT_EXCEEDED,
+            "Too many grace hash join buckets ({} > {}), consider increasing grace_hash_join_max_buckets or max_rows_in_join/max_bytes_in_join",
+            to_size, max_num_buckets);
+    }
+
+    LOG_TRACE(log, "Rehashing from {} to {}", current_size, to_size);
+
+    buckets.reserve(to_size);
+    for (size_t i = current_size; i < to_size; ++i)
+        addBucket(buckets);
+
+    return buckets;
+}
+
+void GraceHashJoin::addBucket(Buckets & destination)
+{
+    BucketPtr new_bucket = std::make_shared<FileBucket>(
+        destination.size(), tmp_data->createStream(left_sample_block), tmp_data->createStream(right_sample_block), log);
+    destination.emplace_back(std::move(new_bucket));
+}
+
+void GraceHashJoin::checkTypesOfKeys(const Block & block) const
+{
+    assert(hash_join);
+    return hash_join->checkTypesOfKeys(block);
+}
+
+void GraceHashJoin::joinBlock(Block & block, std::shared_ptr<ExtraBlock> & not_processed)
+{
+    if (need_left_sample_block.exchange(false))
+    {
+        left_sample_block = block.cloneEmpty();
+        output_sample_block = block.cloneEmpty();
+        hash_join->joinBlock(output_sample_block, not_processed);
+        initBuckets();
```

Comment:
Why do we initialize buckets in so weird place?
AFAIU, first time it's called from `Block JoiningTransform::transformHeader(Block header, const JoinPtr & join)`